오픈마켓 [STEP1] unchain, seohyeon2

seohyeon2 opened the pull request:
안녕하세요 라이언(@ryan-son)! 🦁
언체인(@unchain123)과 현이(@seohyeon2)입니다. 😊
URLSession도, 실제 통신을 하지 않고 Mock 데이터로만 테스트해보는 것도 처음이라 
부족한게 정말정말정말 많습니다..!
라이언이 주실 피드백 받고 다음 스텝 때는 더 나은 코드가 되도록 노력해보겠습니다..!
STEP1 PR 잘부탁드립니다 🙏


## 🥸 코드 설명 

`MockURLSessionDataTask`
- 기존의 URLSessionDataTask와 같은 역할을 하며 MockData를 이용하기 위해 만들어준 클래스입니다.

`MockURLSession`
- URLSession과 같은 기능을 하며 MockData를 받아서 실제 서버와 연결된 것처럼 만들어 주는 클래스 입니다.

`MarketInformation`
- json 데이터의 타입을 정의하였습니다.

`URLSessionProtocol`
- `URLSession`의 `dataTask` 메소드를 교체하기 위해 `dataTask` 랑 동일한 시그니처를 정의해둔 protocol 입니다.

`NetworkManager`
- `URLSession`을 통해 받아온 데이터를 디코딩합니다.

`NetworkError`
- `NetworkManager`의 `fetch` 메서드 시, 발생하는 에러들의 타입입니다.

`ViewController`
- 실제 API의 데이터를 가져오는 `getProductList`메서드가 담겨있습니다.

`OpenMarketTests`
- MockData를 이용하여 테스트를 진행 하였습니다.

<br>

## 😯 궁금한 부분 및 조언 얻고 싶은 부분

### HTTPURLResponse의 사용
```swift
let successResponse = HTTPURLResponse(url: URL(string: url)!,
                                                  statusCode: 200,
                                                  httpVersion: nil,
                                                  headerFields: nil)
```
- 이번에 테스트를 진행하면서 실제 API통신이 아닌 MockData를 이용하여 통신을 하는 상황을 비슷하게 만들어 주어서 테스트를 진행하면서 들었던 의문점입니다. 
- HTTPURLResponse클래스의 사용처가 이렇게 테스트를 하는 용도 뿐인 것인지 아니면 또 다른 용도가 있는지도 궁금합니다. 
- 현업에서도 이 클래스를 이용하여 테스트를 진행하는지도 궁금합니다.

<br>

## 🚀 트러블 슈팅

- `MockURLSessionDataTask` 클래스의 `resumeHandler`를 초기화 시켜줄 때, 아래와 같은 경고가 나옴..
- 아직 해결 못함.. 추후 수정 예정..!

![](https://i.imgur.com/giHpKzx.png)

![](https://i.imgur.com/5j69C21.png)


```swift
init(resumeHandler: @escaping () -> Void) {
        self.resumeHandler = resumeHandler
}
```

unchain123 committed:
feat: JSON파일 추가 및 MarketInformation타입 구현

seohyeon2 committed:
feat: URLSessionProtocol 타입 구현

seohyeon2 committed:
feat: NetworkManager 타입 구현

unchain123 committed:
chore: 컨벤션 수정

unchain123 committed:
refactor: NetworkManager 내부 로직 변경

seohyeon2 committed:
refactor: URLSessionProtocl 타입에 url 매개변수 추가

unchain123 committed:
feat: URLSessionDataTaskProtocol프로토콜 구현

seohyeon2 committed:
feat: MockURLSessionDataTask 타입 구현

unchain123 committed:
feat: MockURLSession타입 구현

seohyeon2 committed:
test: URLSessionTest 코드 작성

unchain123 committed:
feat: NetorkError에러타입 구현

unchain123 committed:
test: test case추가

seohyeon2 committed:
test: Test 실패 케이스 추가

unchain123 committed:
refactor: 불필요한 조건문 제거

seohyeon2 committed:
Merge pull request #1 from seohyeon2/unhyeon_STEP1-2

STEP1-2 PR(Mockdata Test)

seohyeon2 committed:
feat: ViewConreoller에 getProductList 메서드 구현

unchain123 committed:
Merge pull request #2 from seohyeon2/unhyeon_STEP1-2

STEP1-2 PR(feat : getProductList Method)

seohyeon2 committed:
chore: 필요없는 파일 제거

ryan-son commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -37,10 +69,40 @@
+			path = OpenMarketTests;
+			sourceTree = "<group>";
+		};
+		8F1A0BB4287D91B10071EF7A /* Mock */ = {
+			isa = PBXGroup;
+			children = (
+				05B5A7E4287D915B00865293 /* MockURLSessionDataTask.swift */,
+				8F1A0BB5287D91EB0071EF7A /* MockURLSession.swift */,
+			);
+			path = Mock;
+			sourceTree = "<group>";
+		};
```
[P3]
`Mock` 타입들은 테스트를 제외한 환경에서는 사용되지 않을 것 같네요. Unit test 타겟으로 옮기는건 어떨까요?

unchain123 replied:
Mock폴더를 test폴더 내부로 이동해 주었습니다.
[chore: Mock폴더를 test폴더내부로 이동](https://github.com/yagom-academy/ios-open-market/pull/179/commits/d33e41ee7244c10a1469ab8166c98f7c46037b60)

ryan-son commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -165,6 +281,44 @@
+				CODE_SIGN_STYLE = Automatic;
+				CURRENT_PROJECT_VERSION = 1;
+				GENERATE_INFOPLIST_FILE = YES;
+				IPHONEOS_DEPLOYMENT_TARGET = 15.5;
```
[P4]
테스트 타겟도 앱 환경과 동일한 OS 버전으로 맞추는건 어떨까요?

seohyeon2 replied:
앱 환경과 동일하게 수정했습니다!

[test: 테스트 타겟을 앱환경과 동일한 OS로 수정](https://github.com/yagom-academy/ios-open-market/pull/179/commits/c447fbaf051ad2aae71f638890ad49d6510b4e92)

ryan-son commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -165,6 +281,44 @@
+				SWIFT_EMIT_LOC_STRINGS = NO;
+				SWIFT_VERSION = 5.0;
+				TARGETED_DEVICE_FAMILY = "1,2";
+				TEST_HOST = "$(BUILT_PRODUCTS_DIR)/OpenMarket.app/OpenMarket";
```
테스트 타겟의 Host Application이라는 개념에 대해 알아볼까요? 이 테스트 타겟의 Host Application은 어떻게 지정해주는 것이 좋을지 논의해봅시다~! 아래 글부터 알아보기 시작하면 쉽게 시작할 수 있을거에요.

- [XCTest 소요시간 단축하기 - soojin.ro](https://soojin.ro/blog/application-library-test)

unchain123 replied:
라이언이 주신 글을 확인하고 저희 테스트는 Library테스트라는 판단이 들어서 변경을 해주었고 테스트시간이 0.1xx에서 0.01xx로 줄어드는 것을 확인할 수 있었습니다.
[refactor: Application Test를 Library Test로 변경](https://github.com/yagom-academy/ios-open-market/pull/179/commits/40644fbf72a9bc637716ec7676a3c2cc06b85079)

ryan-son replied:
이후에 Host Application이 필요한 테스트가 있다면 타겟을 나누어 테스트를 수행해보는 방법도 생각해볼 수 있을 것 같네요. 고생하셨습니다~ 👏

ryan-son commented on OpenMarket/OpenMarket/Model/MarketInformation.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+struct Page: Decodable {
```
[P3]
과도하게 범용적인 이름은 향후 기능 확장 시 걸림돌이 될 수 있어요. `Page`라는 이름을 가지기 정말 적절한 타입이 등장할 수도 있고, `XXXPage`, `YYYPage`와 같이 접미어로 연속해서 사용되는 경우도 있을 수 있습니다. 최악의 경우에는 First party framework에서 이름을 차지할 수도 있어요. 그 순간이 온다면 이 `Page`가 다른 `Page`들과 어떤 것이 다른지 일일히 설명해주어야 하거나 그제서야 이 타입의 이름을 변경해야 하는 번거로움이 생길 수 있으니 구체적인 이름을 사용해서 문제를 예방하면 좋을 것 같네요.
`Page`라는 이름을 통해서는 어떤 요소인지 파악하기 쉽지 않기도 하니 프로퍼티를 참고해서 이 타입이 어떤 상세한 이름을 가지면 좋을지 생각해보면 좋겠습니다~!

seohyeon2 replied:
네이밍 수정했습니다!

[refactor: MarkInformation안의 구조체 네이밍 수정](https://github.com/yagom-academy/ios-open-market/pull/179/commits/1aec7dd17666cdff812d6f61a60da42048e5749e)

ryan-son commented on OpenMarket/OpenMarket/Model/MarketInformation.swift:
```diff
@@ -0,0 +1,57 @@
+    let createdAt: String
+    let issuedAt: String
+    
+    enum CodingKeys: String, CodingKey {
+        case id
+        case vendorId = "vendor_id"
+        case name
+        case thumbnail
+        case currency
+        case price
+        case bargainPrice = "bargain_price"
+        case dicountedPrice = "discounted_price"
+        case stock
+        case createdAt = "created_at"
+        case issuedAt = "issued_at"
+    }
```
[P5]
프로퍼티 이름과 `CodingKeys`에 정의된 `rawValue` 값이 언더스코어(_) 유무의 차이만 있네요. 아래 내용을 적용하기 좋은 케이스인 것 같아요. 이번 프로젝트에서 적용하지 않아도 좋지만 어떻게 사용할 수 있는지는 학습해보시는 것을 권장합니다~!

- [JSONDecoder.KeyDecodingStrategy.convertFromSnakeCase - Apple Developer](https://developer.apple.com/documentation/foundation/jsondecoder/keydecodingstrategy/convertfromsnakecase)

seohyeon2 replied:
![스크린샷 2022-07-19 오전 12 51 15](https://user-images.githubusercontent.com/50102522/179551895-e8499f96-a362-4ae1-a5fa-0bcdd6fe7b86.png)

주신 링크 참고해서 적용해봤지만 적용이 되지 않았습니다..🥲
저희가 어떤 부분을 놓치고 있을까요?😭

ryan-son replied:
저는 `do-catch`를 통해 `decode(_:from:)`에서 어떤 에러를 던지는지 확인해보았어요.

```
keyNotFound(CodingKeys(stringValue: "dicountedPrice", intValue: nil), Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "pages", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0)], debugDescription: "No value associated with key CodingKeys(stringValue: \"dicountedPrice\", intValue: nil) (\"dicountedPrice\"), converted to dicounted_price.", underlyingError: nil))
```

`dicountedPrice`에 오타가 없는지 확인해볼까요? 😉
옵셔널을 안전하게 처리해주는 것도 좋지만, 에러 문구를 참고할 수 있도록 처리하는 방법도 나쁘지 않다고 생각합니다~

ryan-son commented on OpenMarket/OpenMarket/Model/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+class MockURLSession: URLSessionProtocol {
```
[P5]
이 타입이 더 이상 상속될 필요가 없다고 판단된다면 `final` 키워드를 적용해봅시다~! 이 키워드가 어떤 역할을 하는지, 그리고 적용하면 어떤 이점이 있을 수 있는지 아래 링크들을 통해 확인해보세요~

- [Inheritance - Swift Programming Language Guide](https://docs.swift.org/swift-book/LanguageGuide/Inheritance.html#ID202)
- [Writing High-Performance Swift Code - apple/swift GitHub Repository](https://github.com/apple/swift/blob/main/docs/OptimizationTips.rst#advice-use-final-when-you-know-the-declaration-does-not-need-to-be-overridden)

seohyeon2 replied:
좋은 링크 공유해주셔서 정말 감사합니다!🥺
덕분에 많은 도움이 되었습니다!!

상속할 필요가 없는 파일들 모두 final 처리 해주었습니다!

[refactor: final키워드 추가 및 접근제어자 설정](https://github.com/yagom-academy/ios-open-market/pull/179/commits/edb5c6d4324695281301ebb3f36c0c56ddac294e)

ryan-son commented on OpenMarket/OpenMarket/Model/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+class MockURLSession: URLSessionProtocol {
+    typealias Response = (data: Data?, urlResponse: URLResponse?, error: Error?)
```
`typealias`의 아주 적절한 사용 예시네요 👏. 다음 단계로 `튜플의 typealias`와 `struct` 사이에는 어떤 장단점이 있을지 이야기 나누어보면 좋을 것 같아요.

seohyeon2 replied:
튜플의 typealias와 struct 를 사용하면,
struct에서 데이터 타입을 명시할 때 더 간결하고 가독성 좋은 코드로 작성 할 수 있다는 장점이 있습니다.

단점으로는 각 요소에 대해 접근이 어렵다는 점이 있습니다.
위의 예시에서 튜플의 data 변수에 접근하고자 할 때, Response을 이용하면 접근이 불가했습니다.

ryan-son replied:
지금은 `typealias`를 사용하지 않으시네요. 논의가 어떻게 진행되었나요~?

ryan-son commented on OpenMarket/OpenMarket/Model/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,24 @@
+        self.response = response
+    }
+    
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
```
팀 내부적으로 코딩 컨벤션을 정했나요? 줄바꿈은 어떤 경우에 해주는 것으로 논의되었나요~?

seohyeon2 replied:
Swift Lint를 적용하여 이 기준으로 하기로 결정했습니다😊

ryan-son replied:
line length는 제한을 두지 않고 사용하기로 협의되었나요~?

ryan-son commented on OpenMarket/OpenMarket/Model/Mock/MockURLSessionDataTask.swift:
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+class MockURLSessionDataTask: URLSessionDataTask {
+    private let resumeHandler:() -> Void
```
```suggestion
    private let resumeHandler: () -> Void
```

seohyeon2 replied:
수정했습니다~!

ryan-son commented on OpenMarket/OpenMarket/Model/Mock/MockURLSessionDataTask.swift:
```diff
@@ -0,0 +1,20 @@
+class MockURLSessionDataTask: URLSessionDataTask {
+    private let resumeHandler:() -> Void
+
+    init(resumeHandler: @escaping () -> Void) {
```
이곳에 `@escaping` 어노테이션을 사용해야하는 이유는 무엇이었나요?

unchain123 replied:
여기서 사용하는 resumhandler는 `@escaping`을 붙이지 않아도 escaping클로저로 사용이 되지만 이렇게 어노테이션을 붙여주는 이유는 컴파일러의 최적화때문입니다. non-escaping클로저는 컴파일러가 클로저의 실행이 언제 종료되는지 알기 때문에 때에 따라서  특정 객체에 대한 처리를 생략해 객체의 라이프 사이클을 효율적으로 관리 할 수 있지만 escaping클로저는 함수 밖에서 실행되기 때문에 클로저가 함수 밖에서도 실행되는 것을 보장해야 합니다.
 따라서 escaping클로저에서 사용하는 객체에 대한 추가적인 참조싸이클의 관리를 따로 해줘야 하므로 어노테이션을 사용해서 컴파일러에 명시를 해주는 것이라고 알고 있습니다. 저희가 생각한 부분이 맞을까요??

seohyeon2 replied:
비동기 테스트를 수정하는 과정에서 MockURLSession 관련된 부분도 수정했습니다!

수정 내용은 다음과 같습니다.
기존에 `init` 으로 `resumeHandler:()` 주입해주었지만, Mock테스트에서  이 부분은 불필요할 듯 하여서 삭제했습니다.

[refactor: MockURLSession내부 로직변경](https://github.com/yagom-academy/ios-open-market/pull/179/commits/8ea90f0cedbcf786bf1878a5cd1876692228bb65)

ryan-son replied:
> 여기서 사용하는 resumhandler는 `@escaping`을 붙이지 않아도 escaping클로저로 사용이 되지만 이렇게 어노테이션을 붙여주는 이유는 컴파일러의 최적화때문입니다. non-escaping클로저는 컴파일러가 클로저의 실행이 언제 종료되는지 알기 때문에 때에 따라서 특정 객체에 대한 처리를 생략해 객체의 라이프 사이클을 효율적으로 관리 할 수 있지만 escaping클로저는 함수 밖에서 실행되기 때문에 클로저가 함수 밖에서도 실행되는 것을 보장해야 합니다. 따라서 escaping클로저에서 사용하는 객체에 대한 추가적인 참조싸이클의 관리를 따로 해줘야 하므로 어노테이션을 사용해서 컴파일러에 명시를 해주는 것이라고 알고 있습니다. 저희가 생각한 부분이 맞을까요??

이니셜라이저에서 `@escaping` 어노테이션을 제거하면 컴파일에 실패할거에요. 이니셜라이저에서 resumeHandler를 전달받자마자 호출하는 것이 아니라 이후에 사용하기 위해 프로퍼티에 할당하기 때문이죠. 이 때 `resumeHandler` 코드블럭이 정상적으로 실행되기 위해 자신에게 필요한 요소들을 캡처리스트에 담아 탈출합니다. 이 때문에 Reference Count를 엄격히 관리해줄 필요가 생기는 것이죠. 말씀해주신 것과 같이 Swift의 클로저는 `@escaping` 어노테이션을 작성하지 않은 경우에 기본적으로 즉시 실행되는 클로저라고 판단하지만, 아래와 같이 `@esacping`을 붙이지 않아도 암시적으로 escaping closure라고 판단할 때가 있으니 참고하세요.

```swift
var resumeHandler: (() -> Void)?

init(resumeHander: (() -> Void)?) {
    self.resumeHandler = resumeHandler
}
```

그리고 이후에 활용하기 위해 변수 또는 프로퍼티에 클로저를 담아두는 경우, 그리고 그 클로저가 외부에서 주입되는 경우에는 반드시 escaping closure가 요구되니 알아두시면 좋을 것 같습니다~

ryan-son commented on OpenMarket/OpenMarket/Model/NetworkError.swift:
```diff
@@ -0,0 +1,11 @@
+//  Created by unchain, hyeon2 on 2022/07/12.
+//
+
+enum NetworkError: Error {
+    case outOfRange
+    case failToDecoding
+}
```
따로 에러 메시지는 제공해주지 않을 계획이신지 궁금합니다 😃

unchain123 replied:
원래는 에러 메시지를 얼럿 메시지등에서 처리 할 생각 이었지만 미리 enum타입에 계산프로퍼티로 만들어준 다음 얼럿메시지에서 활용해 보기로 하였습니다.

ryan-son commented on OpenMarket/OpenMarket/Model/URLSessionProtocol.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+protocol URLSessionProtocol {
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
+}
+
+extension URLSession: URLSessionProtocol {}
```
테스트를 위한 모킹 작업 준비가 잘 되었네요 ! 👍

ryan-son commented on OpenMarket/OpenMarket/ViewController.swift:
```diff
@@ -7,12 +7,39 @@
 
 class ViewController: UIViewController {
-
+    
```
의도치 않게 white space가 포함된 것 같네요~ 커밋하실 때 diff를 통해 꼼꼼히 살펴보시면 좋을 것 같아요.

seohyeon2 replied:
넵..! 주의하겠습니다!!!😅

ryan-son commented on OpenMarket/OpenMarket/ViewController.swift:
```diff
@@ -7,12 +7,39 @@
+        let address = "https://market-training.yagom-academy.kr/api/products"
+        let pageNo = URLQueryItem(name: "pageNo", value: "1")
+        let itemsPerPage = URLQueryItem(name: "itemsPerPage", value: "10")
+        let queryItems = [pageNo,itemsPerPage]
```
필요한 주소를 간편하게 만들어주는 타입이 하나 필요하다고 생각하시고 계실 것 같은데요~? 😉

seohyeon2 replied:
주소를 간편하게 만들어주는 타입이라는게 네임스페이스를 모아두는 `Struct` 타입같은 걸 말씀하시는게 맞을까요?😲

ryan-son replied:
`https://market-training.yagom-academy.kr/`라는 base url은 `/api/producs`, `/healthChecker`과 같은 path parameter가 필요할텐데, 이러한 path parameter과 query parameter를 조합해주는 타입을 의미했었습니다~

ryan-son commented on OpenMarket/OpenMarket/ViewController.swift:
```diff
@@ -7,12 +7,39 @@
+                print(error)
+            }
+        }
+     }
```
[P2]
`ViewController`에게 이런 네트워킹 기능을 제공해주는 타입이 하나 있으면 역할을 분리해낼 수 있을 것 같아요. 그럼 테스트하기도 편하지 않을까요?

seohyeon2 replied:
`OpenMarketRequest` 구조체에 해당 기능을 제공해주는 메서드들을 구현했습니다!

[feat: query 및 request 메서드 구현](https://github.com/yagom-academy/ios-open-market/pull/179/commits/cd3bd0a37686c0934c90464ce2c570661f025c87)

[refactor: OpenMarketRequest 메서드 내부 수정](https://github.com/yagom-academy/ios-open-market/pull/179/commits/b3c6141bd27f75f8fad4b3ac869b847643f37baa)

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+import XCTest
+@testable import OpenMarket
+
+class OpenMarketTests: XCTestCase {
```
`Choose Scheme... > Edit Scheme... > Test > Options` 탭에서 `Code Coverage`라는 것을 수집할 수 있어요. `Code Coverage`는 테스트를 통해 검증된 코드의 비율을 의미합니다. 이 옵션을 활성화하고 테스트를 수행한 다음 Report Navigator > 최근 수행한 Test의 Coverage를 확인하셔서 여러분이 작성한 테스트가 얼마나 여러분의 코드를 검증했는지 확인해보시는 것도 재미있을 거에요. 어떤 부분의 테스트를 더 추가해볼 수 있을지 이야기 나누어보면 좋을 것 같습니다~!

<img width="472" alt="image" src="https://user-images.githubusercontent.com/69730931/178791867-7196d855-b875-47ce-b306-35bdb6af110e.png">

<img width="934" alt="image" src="https://user-images.githubusercontent.com/69730931/178791947-2f32e05c-23ab-4211-b846-bf283c041746.png">

<img width="898" alt="image" src="https://user-images.githubusercontent.com/69730931/178792375-b7d187cc-447a-4cfb-9391-3a4cf5241445.png">

unchain123 replied:
<img width="500" alt="스크린샷 2022-07-14 오후 8 52 14" src="https://user-images.githubusercontent.com/98506825/178976513-d639b30a-acc4-466e-9111-60b236011511.png">

- 말씀하신대로 커버리지를 확인해 봤습니다. 처음 목적이었던 목데이터에 관한 테스트는 100%였기 때문에 초기의 목적은 달성 했다고 보여 집니다. 뷰컨트롤러에 관한 테스트는 향후에 테스트할 기능이나 네트워크 통신 기능을 넣게 된다면 추가하도록 하겠습니다.

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+        sut = nil
+    }
+    
+    func test_fetchData_Data가_있고_statusCode가_200일때() {
```
[P3]
Given, when, then 형식으로 작성해두었으니 해당 내용들을 간추려서 테스트 이름을 다시 지어보면 좋을 것 같아요.
`{when}(일 때/인 경우) {then}이다`와 같은 형식으로 작성하면 어떤 경우에 어떤 원하는 결과를 얻을 수 있는지 설명하기 좋을 것 같아요. `sut`이 `NetworkManager`라면 테스트 메서드의 이름을 모아두었을 때 `NetworkManager`가 수행할 수 있는 기능 명세가 되는 형태로 작성하면 좋아요.

seohyeon2 replied:
네이밍 수정했습니다!

[test: test명 변경](https://github.com/yagom-academy/ios-open-market/pull/179/commits/98cc60789f00512adf01063f828c44cf64c7a62c)

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+import XCTest
+@testable import OpenMarket
+
+class OpenMarketTests: XCTestCase {
```
[P3]
테스트 타입 이름이 프로젝트 이름과 동일해서 넓은 범위를 커버해야할 것 같아요. 테스트 대상에 맞추어 타입 이름을 변경해보는 것도 좋겠네요.

seohyeon2 replied:
네이밍 수정했습니다!

[test: test명 변경](https://github.com/yagom-academy/ios-open-market/pull/179/commits/98cc60789f00512adf01063f828c44cf64c7a62c)

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+        // given
+        let url = "https://market-training.yagom-academy.kr/"
+        let mockResponse: MockURLSession.Response = {
+            let data = NSDataAsset(name: "products", bundle: .main)?.data
```
여기에서 `bundle`이 의미하는 것은 무엇인가요? `NSDataAsset(name: "products")`로 작성하면 동작할까요? `bundle`이 커버하는 영역은 어디까지일지 알아보면 좋을 것 같아요. 분명히 이후에 큰 도움이 될거에요. 추가로, Swift Package에서 리소스에 접근하기 위해서 어떤 번들에 접근해야하는지 알아보는 것도 재미있을거에요.

- [NSDataAsset - Apple Developer](https://developer.apple.com/documentation/uikit/nsdataasset/1403436-init)
- [Bundle - Apple Developer](https://developer.apple.com/documentation/foundation/bundle)
- [Bundling Resources with a Swift Package - Apple Developer](https://developer.apple.com/documentation/xcode/bundling-resources-with-a-swift-package)

seohyeon2 replied:
이 코드에서 `bundle`이 의미하는 것은  `products`라는 이름의 데이터를 저장하는 데 사용되는 번들입니다.
`NSDataAsset(name: "products")` 라고 작성하면 데이터 에셋을 찾을 수 없어 동작하지 못합니다.

그래서 `NSDataAsset.init(name: "products", bundle: Bundle(for: NetworkManagerTests.self))` 로 수정했습니다.

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+                result = product
+                
+                // then
+                guard let data = NSDataAsset(name: "products")?.data else {
```
[P3]
"GET", "products" 처럼 literal string을 반복해서 사용하면 언젠가는 실수할 수 있어요. 이런 값들을 안전하게 사용할 수 있는 방법을 구상해보고, 저와 이야기 나누어봐요~

seohyeon2 replied:
자주 쓸 literal은 네임스페이스를 따로 마련하여 수정하였습니다!
테스트파일의 `"products"` 경우, 테스트를 위한 json 파일이기 때문에 재사용할 일이 없다 생각하여 네임스페이스로 빼지 않았습니다.😌

[feat: NetWorkNamespace 타입 구현](https://github.com/yagom-academy/ios-open-market/pull/179/commits/da8065819c1a31cea800faed719c0e367d5e8f1b)

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+        // given
+        struct MarketInformationTest: Decodable {
+            let pageNo: Int
+        }
```
[P1]
이 테스트는 실패하네요. 어떤 것이 잘못되었는지 `products.json` 파일을 보시면 바로 아실 수 있으실 것 같습니다!

seohyeon2 replied:
비동기 테스트 모두 수정하였습니다!
새롭게 추가한 테스트 다시 봐주시면 감사하겠습니다! 🙇🏻‍♀️

[test: 상품리스트 읽어와서 매핑에 성공하는 테스트 생성](https://github.com/yagom-academy/ios-open-market/pull/179/commits/7ef78b8cc75c357dcda40c8b755a77454555ddd6)
[test: Asset 파일 읽어오는 코드 수정](https://github.com/yagom-academy/ios-open-market/pull/179/commits/13cdad2888e171f34e4d5e0f9c4e50209191bcab)
[test: 상품리스트 읽어와서 매핑에 실패하는 테스트 생성](https://github.com/yagom-academy/ios-open-market/pull/179/commits/3e29ecd2024ed763e4e81138dcaabcdf87e2ae24)

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+        }
+        
+        
+        let url = "https://market-training.yagom-academy.kr/"
```
상품 목록을 받기 위해서는 적절한 base URL에 쿼리 파라미터를 추가해서 요청해야 할텐데요, 그러면 주소도 테스트 대상에 포함해야 할까요? `URLRequest`는요? 여러분은 어떻게 생각하세요? 🤔

seohyeon2 replied:
`주소`와 `URLRequest` 모두  테스트 대상이 아니라고 생각합니다..!

Mock 데이터를 만들어서 테스트하게 되면 요청과 응답을 조작할 수 있다고 알고 있습니다.
테스트 시 요청과 응답을 조작하는 것은 통신이 잘 되는지에 대한 것보다,
네트워크에서 받아온 데이터 기준으로 우리가 작성한 코드가 예상대로 동작하는지를 테스트 하기 위함이라고 생각했습니다.
응답이 왔다고 가정한 상태에서 그 값을 의도한 대로 잘 처리하는지 테스트하는 것이므로 
`주소`와 `URLRequest` 모두  테스트 대상이 아니라고 느꼈습니다.

때문에 테스트파일에서 주소는 제외해주었습니다😌

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+                                                  headerFields: nil)
+            return (data: data,
+                    urlResponse: successResponse,
+                    error: nil)
```
변경하라는 의미는 아니지만, `typealias`도 아래와 같은 방식으로 작성할 수 있으니 참고하세요~ 인스턴스를 생성할 수 있는 일반적인 타입과 굉장히 비슷하지 않나요? 😊

```swift
MockURLSession.Response(data, successResponse, nil)
MockURLSession.Response(
    data: data, 
    urlResponse: successResponse,
    error: nil
)
```

ryan-son commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,134 @@
+                result = error
+                XCTAssertNotNil(result)
+            }
+        }
```
[P4]
`Result` 타입의 결과가 `success`인 경우 테스트가 실패하도록 구성해보는건 어떨까요?

seohyeon2 replied:
비동기 테스트 모두 수정하였습니다!
새롭게 추가한 테스트 다시 봐주시면 감사하겠습니다! 🙇🏻‍♀️

[test: 상품리스트 읽어와서 매핑에 실패하는 테스트 생성](https://github.com/yagom-academy/ios-open-market/pull/179/commits/3e29ecd2024ed763e4e81138dcaabcdf87e2ae24)

ryan-son reviewed (changes requested):
현이, 언체인! Step 1 수행하시느라 고생 많으셨어요. 실제 네트워크 작업을 수행하지 않고 로컬에서 네트워킹 테스트를 하기 쉽지 않으셨을텐데 과정을 빠르게 끝내셨네요~! 👍

여러분들이 질문하셨던 부분부터 짚고가겠습니다~

# 질의응답

>### HTTPURLResponse의 사용
```swift
let successResponse = HTTPURLResponse(
    url: URL(string: url)!,
    statusCode: 200,
    httpVersion: nil,
    headerFields: nil
)
```
>- HTTPURLResponse클래스의 사용처가 이렇게 테스트를 하는 용도 뿐인 것인지 아니면 또 다른 용도가 있는지도 궁금합니다. 
>- 현업에서도 이 클래스를 이용하여 테스트를 진행하는지도 궁금합니다.

여러분들이 `NetworkManager` 타입을 구현하시면서 이미 `HTTPURLResponse`로 타입캐스팅한 경험을 가지고 계실거에요. HTTP 상태 코드가 어떤 범주에 있는지 파악하기 위해 사용하셨었죠. 이처럼 `HTTPURLResponse` 자체는 HTTP 통신 요청을 통해 전달받은 응답을 의미하기 때문에 테스트가 목적이 아니라면 모바일과 같은 클라이언트 환경에서 직접 만드는 경우는 없다고 보셔도 됩니다. 
현업과 관계없이 로컬에서 네트워크 응답과 관련된 테스트를 하려면 `HTTPURLResponse` 혹은 그에 준하는 타입을 정의하여 사용해야 합니다. 더 자세한 내용을 알아보시고 싶으시다면 `Moya`라는 라이브러리가 로컬 환경에서 테스트를 수행하기 위해 어떤 인터페이스를 제공하는지, 그 인터페이스는 어떻게 구현되어 있는지 살펴보시면 궁금한 부분이 많이 해소되실 것 같네요. 만약 내용이 너무 많다고 생각되시면 제가 도움을 드릴 수 있으니 언제든지 찾아주세요~

- [Moya - GitHub Repository](https://github.com/Moya/Moya)
- [Testing - Moya](https://github.com/Moya/Moya/blob/master/docs/Testing.md)
- [`EndpointSampleResponse` - Moya](https://github.com/Moya/Moya/blob/42b4da6cf00ce35874f2aebebefb9564c2fdb799/Sources/Moya/Endpoint.swift#L4-L14)
- [`createStubFunction(_:forTarget:withCompletion:endpoint:plugins:request:)` - MoyaProvider+Internal.swift](https://github.com/Moya/Moya/blob/42b4da6cf00ce35874f2aebebefb9564c2fdb799/Sources/Moya/MoyaProvider%2BInternal.swift#L106-L143)
- [`Moya.Response` - Moya](https://github.com/Moya/Moya/blob/42b4da6cf00ce35874f2aebebefb9564c2fdb799/Sources/Moya/Response.swift#L4-L38)

# `URLSessionDataTask` 이니셜라이저의 대안
PR을 통해 말씀해주셨듯이 이니셜라이저가 deprecated 되어버려 곤란하실거에요. Apple에서 소개한 대안이 궁금하시다면 싶으시다면 아래 영상을 참고하세요~ 네트워킹과 관련된 내용만 보실 계획이시라면 13분 40초 정도까지만 보시면 될거에요.
- [Testing Tips & Tricks - WWDC18](https://developer.apple.com/videos/play/wwdc2018/417/)

# 개별 피드백 P 태그 안내
각 코멘트에는 P1 ~ P5까지의 태그가 포함되어있습니다. 태그의 의미는 아래와 같으니 참고하세요. [뱅크샐러드 개발 문화](https://blog.banksalad.com/tech/banksalad-code-review-culture/)에서 차용하였습니다.

<img width="786" alt="image" src="https://user-images.githubusercontent.com/69730931/176375527-04436d3c-bb00-4abd-a1d9-224b1c65b9eb.png">

계속해서 화이팅입니다~ 💪 도움이 필요하시면 언제든지 말씀해주세요~

unchain123 committed:
chore: Mock폴더를 test폴더내부로 이동

seohyeon2 committed:
test: 테스트 타겟을 앱환경과 동일한 OS로 수정

unchain123 committed:
refactor: Application Test를 Library Test로 변경

seohyeon2 committed:
refactor: MarkInformation안의 구조체 네이밍 수정

unchain123 committed:
refactor: final키워드 추가 및 접근제어자 설정

seohyeon2 committed:
style: Swift Lint 적용에 따른 수정

seohyeon2 committed:
chore: gitignore 수정

seohyeon2 committed:
style: 화이트스페이스 제거 및 Swift Lint 추가

seohyeon2 committed:
chore: 누락된 파일 추가

unchain123 committed:
refactor: 에러문구 추가

unchain123 committed:
test: test명 변경

unchain123 committed:
merge: refactoring 브랜치 머지

# Conflicts:
#	.gitignore
#	OpenMarket/.swiftlint.yml
#	OpenMarket/OpenMarket.xcodeproj/project.pbxproj

seohyeon2 committed:
feat: NetWorkNamespace 타입 구현

seohyeon2 committed:
docs: README 수정

unchain123 committed:
refactor: fetch메서드 기능분리

unchain123 committed:
test: 테스트 항목 삭제

unchain123 committed:
refactor: fetch메서드 구조변경

seohyeon2 committed:
refactor: getProductInquiry 메서드 위치 및 구조 변경

unchain123 committed:
refactor: MockURLSession내부 로직변경

seohyeon2 committed:
feat: query 및 request 메서드 구현

unchain123 committed:
refactor: OpenMarketRequest 메서드 내부 수정

unchain123 committed:
feat: getProductList메서드 생성

seohyeon2 committed:
feat: requestProductDetail 메서드 구현

unchain123 committed:
test: 상품리스트 읽어와서 매핑에 성공하는 테스트 생성

seohyeon2 committed:
test: Asset 파일 읽어오는 코드 수정

seohyeon2 committed:
test: 상품리스트 읽어와서 매핑에 실패하는 테스트 생성

unchain123 committed:
feat: ModelNameSpace파일 생성

unchain123 committed:
refactor: NetworkNamespace에서 ModelNameSpace로 내용분리

ryan-son commented on .gitignore:
```diff
@@ -1,53 +1,21 @@
+# CocoaPods - Only use to conserve bandwidth / Save time on Pushing
+#           - Also handy if you have a large number of dependant pods
+#           - AS PER https://guides.cocoapods.org/using/using-cocoapods.html NEVER IGNORE THE LOCK FILE
+Podfile
+Podfile.lock
```
`Podfile`과 `Podfile.lock`은 어떤 역할을 가지고 있을까요? `gitignore`에 추가하면 어떤 영향이 있을까요?

ryan-son replied:
최근에는 Swift Package Manager의 안정성이 보완되어 체감상 신규 프로젝트에서는 CocoaPods를 사용하는 빈도가 줄고 있는 추세이니 린팅은 Homebrew를 통해 설치해서 사용하는 방안도 고려해볼 수 있을 것 같아요.

ryan-son commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,15 @@
+    # 라인 뒤에 공백이 없어야 합니다.
+    - trailing_whitespace
+
+    # 강제 캐스팅은 피해야합니다.
+    - force_cast
```
강제 캐스팅을 제외할 룰에 추가하신 이유가 있을까요?

ryan-son commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,15 @@
+    # 강제 캐스팅은 피해야합니다.
+    - force_cast
+
+    - line_length
```
[SwiftLInt Documentation](https://realm.github.io/SwiftLint/index.html)에서 규칙들의 상세사항을 확인하실 수 있습니다.
[line_length](https://realm.github.io/SwiftLint/line_length.html)의 기본 구성은 아래와 같이 정의되어 있네요.

```
warning: 120
error: 200
ignores urls: false
ignores function declarations: false
ignores comments: false
ignores interpolated strings: false
```

기본 구성을 변경하고 싶으시다면 아래와 같이 작성하실 수 있습니다.
```
line_length:
  warning: 200
  error: 300
  ignores_urls: true
```

ryan-son commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,15 @@
+
+excluded: # SwiftLint 검사에서 제외할 파일 경로
+  - Pods
+  - OpenMarket/AppDelegate.swift
+  - OpenMarket/SceneDelegate.swift
```
[P4]
검사 대상에서 제외하기보다 다른 파일들과 형식을 동일하게 맞추어주는건 어떨까요? 불필요한 주석을 제거하고, 메서드의 형식을 다른 타입들의 메서드와 동일하게 만들어주면 될 것 같아요.

ryan-son commented on OpenMarket/OpenMarket/Model/NameSpace/ModelNameSpace.swift:
```diff
@@ -0,0 +1,74 @@
+//  Created by unchain, hyeon2 on 2022/07/19.
+//
+
+enum ModelNameSpace {
```
[P4]
이 이름으로는 목적이 분명해보이지 않을 수 있으니 요청 쿼리에 사용되는 네임스페이스라는 의미를 추가해보는건 어떨까요?

ryan-son commented on OpenMarket/OpenMarket/Model/NameSpace/NetworkNamespace.swift:
```diff
@@ -0,0 +1,29 @@
+            return "PATCH"
+        case .del:
+            return "DEL"
+        case .url:
+            return "https://market-training.yagom-academy.kr/api/products"
```
[P3]
상품리스트에 한정한 URL이 되겠네요. 이 부분은 base url과 path parameter를 받아 조합할 수 있도록 확장성 있게 구성해보는건 어떨까요?

ryan-son commented on OpenMarket/OpenMarket/Network/OpenMarketRequest.swift:
```diff
@@ -0,0 +1,42 @@
+struct OpenMarketRequest {
+    private let address = NetworkNamespace.url.name
+    
+    func createQuery(of pageNo: String = "1", with itemsPerPage: String = "20") -> [URLQueryItem] {
+        let pageNo = URLQueryItem(name: ModelNameSpace.pageNo.name, value: pageNo)
+        let itemsPerPage = URLQueryItem(name: ModelNameSpace.itemsPerPage.name, value: itemsPerPage)
```
[P4]
이 메서드를 사용하는 입장에서는 페이지 번호와 페이지당 상품 개수가 `String` 타입이라는 점에 대해 의문을 가질 수도 있을 것 같네요. 사용성을 고려해서 정수 타입으로 입력을 받는건 어떨까요?

ryan-son commented on OpenMarket/OpenMarket/ViewController.swift:
```diff
@@ -7,12 +7,26 @@
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        getProductList(about: "3", "1")
```
[P3]
호출하는 쪽에서는 about과 두번째 매개변수가 어떤 의미인지 알아보기 어려울 수 있을 것 같네요. 특별한 이유가 없다면 매개변수가 가지는 의미를 그대로 노출해주어도 좋을 것 같아요.

ryan-son commented on OpenMarket/OpenMarketTests/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,44 @@
+//  Created by unchain, hyeon2 on 2022/07/12.
+//
+
+import Foundation
+import UIKit
```
`UIKit`에 `Foundation` 프레임워크가 포함되어 있으니 `UIKit`만 작성해주어도 작동할거에요. `NSDataAsset`이라는 클래스만 import하시기를 원하신다면 아래와 같은 방식으로 작성해보실 수도 있지만, 어차피 이 모듈에서 `UIKit` 전체를 사용하고 있으니 `UIKit` 자체를 import 하는 것과 아무런 차이가 없겠네요.

```swift
import class UIKit.NSDataAsset
```

ryan-son commented on OpenMarket/OpenMarketTests/Mock/MockURLSessionDataTask.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+final class MockURLSessionDataTask: URLSessionDataTask {
+    var resumeHandler: () -> Void = {}
```
`() -> Void` 타입의  `{}` 클로저와 `Void` 타입의 `{}()`는 각각 어떤 의미를 가지고 있을까요?

ryan-son commented on OpenMarket/OpenMarketTests/NetworkManagerTests.swift:
```diff
@@ -0,0 +1,75 @@
+        mockSession = nil
+    }
+
+    func test_상품리스트를_요청하면_매핑에_성공한다() {
+        // given
+        guard let assetData = NSDataAsset.init(name: "products", bundle: Bundle(for: NetworkManagerTests.self)) else {
+            XCTFail("asset fail")
+            return
+        }
+        let promise = expectation(description: "매핑 테스트")
+
+        // when
+        let result = try? JSONDecoder().decode(MarketInformation.self, from: assetData.data)
+        sut?.getProductInquiry(request: nil, completion: { expectation in
+            switch expectation {
+            case .failure(_):
+                XCTFail("failure")
+            case .success(let data):
+                guard let expectation = try? JSONDecoder().decode(MarketInformation.self, from: data) else {
+                    XCTFail("decode error")
+                    return
+                }
+                XCTAssertEqual(result?.pageNo, expectation.pageNo)
+                promise.fulfill()
+            }
+        })
+        
+        wait(for: [promise], timeout: 5)
+    }
+
+    func test_상품리스트를_요청하면_매핑에_실패한다() {
```
[P3]
어떤 조건일 때 성공하는지, 그리고 실패하는지가 메서드 이름에 포함되어 있지 않아 혼란스울 수 있겠네요. 성공할 때와 실패할 때 어떤 조건이 다른가요?

ryan-son replied:
이 테스트는 실패하네요. 어떤 의도였는지 이따 이야기 나누어봅시다~ 😊

ryan-son reviewed (approved):
언체인, 현이! 고생 많으셨어요.

이따 회의실에서 이야기 나누어보고 의문점들 해소해볼 수 있는 시간 가지도록 합시다~
